### PR TITLE
Remove `'static` requirement on the inner `Action` object

### PR DESCRIPTION
**Pull request by rlycx**
Removes the `'static` requirement on the `dyn Action` in `Inner::Eval`.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -78,11 +71,12 @@ impl IO<()> {
             o: &mut std::io::stdout(),
             e: &mut std::io::stderr(),
         })
+
     }
 
     /// Perform the IO action and return anything printed to stdout
     /// as a string.
-    pub fn capture (IO (inner): IO<()>) -> Result<String> {
+    pub fn capture (IO (inner): IO<'_>) -> Result<String> {
```
Comment (MaeIsBad) [inline src/lib.rs:134]: Run has a comment mentioning that it's a static assoc function, maybe add it here too?

File: src/lib.rs [hunk 2]
```diff
@@ -1,23 +1,23 @@
...
+use std::{ io::{ Read, Result, Write }, path::Path, mem, fs, marker::PhantomData };
 use Inner::{ Eval, Pure, Done };
 
 /// A chain of lazy, composable IO actions.
-pub struct IO <O> (Inner<O>);
+pub struct IO <'a, O = ()> (Inner<'a, O>);
 
-enum Inner <O> {
-    Eval (Box<dyn Action<Output = O>>),
+enum Inner <'a, O> {
```
Comment (MaeIsBad) [inline src/lib.rs]: Consider commenting what each state of inner actually means. It might be too verbose for someone and an equivalent of a "declare the variable i" comment but it wasn't 100% obvious for me

File: src/lib.rs [hunk 3]
```diff
@@ -1,23 +1,23 @@
...
+    Eval (Box<dyn Action<'a, Output = O> + 'a>),
```
Comment (MaeIsBad) [inline src/lib.rs]: Leave a comment explaining why you chose to use box here? 

My understanding is it would make the types hellishly complicated but would it be possible to maybe instead do something similar to warp filters and allow the user to box thing easily without making it explicitly necessary?

Comment (MaeIsBad) [reply src/lib.rs]: I think not wanting to overcomplicate code by messing with rust's generics is a valid reason, I don't think the Box<dyn> here affects performance in any significant way for most projects, especially considering how expensive most IO already is[citation needed]

File: src/lib.rs [hunk 4]
```diff
@@ -1,23 +1,23 @@
...
 
 /// Primitive for [`IO`]. This is the "unsafe", lower level interface that can be
 /// abused (if it weren't for other safety precautions such as the `Context` being
 /// private).
-trait Action {
+trait Action<'a> {
     /// The value the action will evaluate to.
-    type Output;
+    type Output: 'a;
     /// Attempts to perform an [`Action`] more than once is generally cause for
```
Comment (MaeIsBad) [inline src/lib.rs]: The world "generally" is imprecise, I would remove it from here(or replace it with the word always) considering that as far as I can see calling it more than once will always cause a panic

File: src/lib.rs [hunk 5]
```diff
@@ -30,17 +30,10 @@ trait Action {
     fn perform (&mut self, cx: Context<'_>) -> Result<Self::Output>;
```
Comment (MaeIsBad) [inline src/lib.rs:89]: Would having a non-panicking try_perform that returns an error when called more than once be a good idea?

File: src/lib.rs [hunk 6]
```diff
@@ -105,23 +99,64 @@ impl IO<()> {
 }
 
 /// Various combinators on the IO monad.
-impl<O> IO<O> {
+impl<'a, O> IO<'a, O>
```
Comment (MaeIsBad) [inline src/lib.rs]: Would it make sense to move this into a separate module(and file)?

File: src/lib.rs [hunk 7]
```diff
@@ -105,23 +99,64 @@ impl IO<()> {
...
+
+        IO::new(FromFn (Some (f), PhantomData))
+    }
 
     /// Perform a monadic binding operation. For the IO monad, this
     /// means simply sequencing `self` to happen before calling `f`
     /// on the result and then performing the value returned from
     /// `f`.
-    pub fn bind <R> (self, f: impl FnOnce (O) -> IO<R> + 'static) -> IO<R>
+    pub fn bind <R> (self, f: impl FnOnce (O) -> IO<'a, R> + 'a) -> IO<'a, R>
```
Comment (MaeIsBad) [inline src/lib.rs]: Can you use simpler language here? As a rust dev I have no idea what "sequencing self" could mean ><

File: src/lib.rs [hunk 8]
```diff
@@ -105,23 +99,64 @@ impl IO<()> {
...
-        O: 'static,
-        R: 'static,
+        R: 'a,
     {
         // Declare a new struct that holds the captured values
-        struct Bind <A, F> (IO<A>, Option<F>);
+        struct Bind <'c, A, F> (IO<'c, A>, Option<F>);
 
-        impl<A, F, X> Action for Bind<A, F>
+        impl<'c, 'd: 'c, A, F, X> Action<'d> for Bind<'c, A, F>
```
Comment (MaeIsBad) [inline src/lib.rs]: I have never done this in a project before but maybe using longer more descriptive names would make this more readable? I've seen some rust devs advocate for this before but I haven't really looked into if it actually made the code more readable

File: src/lib.rs [hunk 9]
```diff
@@ -153,7 +188,7 @@ impl<O> IO<O> {
     }
 
     /// Apply a function to the wrapped value.
-    pub fn map <R> (self, f: impl FnOnce (O) -> R + 'static) -> IO<R>
+    pub fn map <R> (self, f: impl FnOnce (O) -> R + 'a) -> IO<'a, R>
     where
         O: 'static,
         R: 'static,
```
Comment (MaeIsBad) [inline src/lib.rs]: Is O and R being static here necessary?

File: src/lib.rs [hunk 10]
```diff
@@ -212,56 +242,62 @@ impl<O> IO<O> {
...
 /// # Ok (()) }
 /// ```
-pub fn pure <T> (value: T) -> IO<T> {
+pub fn pure <'a, T> (value: T) -> IO<'a, T> where T: 'a {
     IO (Pure (value))
 }
 
 /// Ignore the output value of `x`.
-pub fn void <T> (x: IO<T>) -> IO<()>
+pub fn void <'a, T> (x: IO<'a, T>) -> IO<'a>
```
Comment (MaeIsBad) [inline src/lib.rs]: I think the docs for this function could be improved? As a non FP dev this makes me guess that it either: "Does this make IO evaluate? Or is this equivalent to `.map(|_| ())`?" I think the second one is right but I don't really know

File: src/lib.rs [hunk 11]
```diff
@@ -212,56 +242,62 @@ impl<O> IO<O> {
...
-    impl Action for PrintLn {
+    impl Action<'static> for PrintLn {
         type Output = ();
         fn perform (&mut self, Context { o, .. }: Context<'_>) -> Result<Self::Output> {
             writeln!(o, "{}", self.0)
         }
     }
-    
+
     IO::new(PrintLn (s.to_string()))
```
Comment (MaeIsBad) [inline src/lib.rs]: Could you avoid having the IO take ownership of `s` by making it's lifetime not static?

File: src/lib.rs [hunk 12]
```diff
@@ -212,56 +242,62 @@ impl<O> IO<O> {
...
 }
 
 /// Ignore the output value of `x`.
-pub fn void <T> (x: IO<T>) -> IO<()>
+pub fn void <'a, T> (x: IO<'a, T>) -> IO<'a>
 where
     T: 'static
 {
     x.replace(())
 }
```
Comment (MaeIsBad) [inline src/lib.rs]: I feel like the various IO primitives(? I think that's what you call them) belong in a different module. If you wanna get fancy with it I think the rest of the crate could do just fine only relying on `alloc`, so you could make `std` an optional dependency

File: src/lib.rs [hunk 13]
```diff
@@ -212,56 +242,62 @@ impl<O> IO<O> {
...
+        Ok (s)
     })
 }
 
 /// Acquire a resource and apply a function to it.
-pub fn bracket <I, O, F> (
-    acquire: IO<I>,
-    release: IO<()>,
-) -> impl FnOnce (F) -> IO<O>
+pub fn bracket <'a, I, O, F> (
```
Comment (MaeIsBad) [inline src/lib.rs]: I'm struggling to understand what this function does, could you explain it in layman's terms?

File: src/lib.rs [hunk 14]
```diff
@@ -288,6 +324,37 @@ where
...
     |f| IO::new(Bracket (acquire, release, Some (f)))
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    /// Test whether [`bracket`] sequences actions correctly.
+    fn sequencing_bracket () {
```
Comment (MaeIsBad) [inline src/lib.rs]: Would a name like `bracket_sequences_actions_correctly` be better here? 
I'm basing this entirely off of cargo naming the default test function when you generate a project with `cargo new --lib` "`it_works`"

Comment (MaeIsBad) [reply src/lib.rs]: That seems to be the test naming conventions that the rust stdlib sometimes follows(but it's inconsistent about it)

**Review by MaeIsBad — COMMENTED** [review]
I think a few docs with general code samples that show off the entire library would be nice to add, in general most of my issues come from wanting docs that are more FP beginner friendly, which could be nice considering that's most of the rust dev community

File: src/lib.rs (hunk 4 above)
Comment (rlycx) [reply src/lib.rs]: In the case of `println` the action is repeatable, but yeah, it still *shouldn't* be repeated

File: src/lib.rs (hunk 5 above)
Comment (rlycx) [reply src/lib.rs:89]: I was thinking something similar before, but decided against for the following reasons:

1. I couldn't find a good `std::io::ErrorKind` to fit and I'd prefer to just use `std::io::Result` for simplicity (also since these are IO actions, it makes sense to return only IO errors imo)
2. Outside of gross `unsafe` abuse, there should be no way to repeat an `Action` due to `IO`'s interface preventing it
3. End users can't actually call this function or provide its `Context` argument so there is no way for a user to mistakenly or willingly trigger a panic due to the requirement being violated

File: src/lib.rs (hunk 6 above)
Comment (rlycx) [reply src/lib.rs]: Yeah, will do

File: src/lib.rs (hunk 3 above)
Comment (rlycx) [reply src/lib.rs]: I'd also like to avoid exposing the `Action` trait to the end user because that would make it more difficult to keep the internals as internals. Also, indeed, I want to avoid the psychic damage of introducing *more* parameters to end users (at the cost of redirecting that psychic damage to contributors)

File: src/lib.rs (hunk 9 above)
Comment (rlycx) [reply src/lib.rs]: No, will fix

File: src/lib.rs (hunk 13 above)
Comment (rlycx) [reply src/lib.rs]: @MaeIsBad I have updated the documentation, feedback would be appreciated!

File: src/lib.rs (hunk 7 above)
Comment (rlycx) [reply src/lib.rs]: I've added some additional explanation, let me know if it helps :)

File: src/lib.rs [hunk 15]
```diff
@@ -334,7 +431,7 @@ mod tests {
 
     #[test]
     /// Test whether [`bracket`] sequences actions correctly.
```
Comment (MaeIsBad) [inline src/lib.rs:216]: This comment seems redundant now, it's basically the same as the test function name

File: src/lib.rs (hunk 10 above)
Comment (MaeIsBad) [reply src/lib.rs]: These docs are so great, clean and understandable, way better what I had in mind 👍. I love them

File: src/lib.rs [hunk 16]
```diff
@@ -242,30 +271,74 @@ where
...
+/// // The action needs to return `()` so we just print the
+/// // value
+/// let action = f("meow").bind(println);
+/// 
+/// // Run the action
+/// let result: usize = IO::capture(action)?.trim().parse().unwrap();
+///
+/// assert_eq!(result, 0);
+/// # Ok (()) }
+/// ```
```
Comment (MaeIsBad) [inline src/lib.rs]: Good girl, I really like these docs ❤️

File: src/lib.rs (hunk 7 above)
Comment (MaeIsBad) [reply src/lib.rs]: This might be out of scope for the docs here, but I think maybe showing the function signature for map may be useful as a reminded/explanation of what the difference between map and bind/and_then is? The docs for this function are already aimed at a "FP baby" like me, since anyone familiar with FP will already know what bind is.

Other than that the and_then comparison is really good and completely clears up what this does completely.

File: src/lib.rs (hunk 2 above)
Comment (MaeIsBad) [reply src/lib.rs:10]: Expended is a hard word. It's not that complicated,but at least for me it's not in the "L1 brain word cache" which does make reading it a bit less smooth, I like the comments and don't really have an opinion on Eval -> Chain but I think end is a better name than expanded

This may be my 1am brain speaking

**Review by MaeIsBad — COMMENTED** [review]
1am code review let's gooooo

File: src/lib.rs (hunk 16 above)
Comment (rlycx) [reply src/lib.rs]: one hundred emoji

File: src/lib.rs (hunk 2 above)
Comment (rlycx) [reply src/lib.rs]: What about `Depleted`?

Comment (MaeIsBad) [reply src/lib.rs]: Doesn't feel like it's any different for me tbh?

File: src/lib.rs (hunk 15 above)
Comment (rlycx) [reply src/lib.rs:216]: I still like having the comment there tbh

File: src/lib.rs (hunk 8 above)
Comment (RileyApeldoorn) [reply src/lib.rs]: Coming up with real names is hard though :sob:

Do you have any suggestions? Otherwise I think I'm gonna leave it like this lol

Comment (rlycx) [reply src/lib.rs]: I've changed the implementation so it doesn't need all these parameters :D

File: src/lib.rs (hunk 13 above)
Comment (MaeIsBad) [reply src/lib.rs]: Looks great 👍

File: src/lib.rs (hunk 2 above)
Comment (MaeIsBad) [reply src/lib.rs]: After IRL discussion we agreed to keep Depleted as the name, The descriptiveness is better than using simple words

File: src/combinator.rs [hunk 17]
```diff
@@ -0,0 +1,304 @@
...
+    /// You can think of this as the [`IO`] version of [`Option`]'s
+    /// [`and_then`](Option::and_then) function, except fully lazy.
+    ///
+    /// If we simplify the signatures somewhat, the similarities are
+    /// clear(er) to see:
+    ///
+    /// ```text
+    /// fn and_then <F, U> (self: Option<T>, f: F) -> Option<U>
+    /// where
+    ///     F: FnOnce (T) -> Option<U>;
```
Comment (MaeIsBad) [inline src/combinator.rs:112]: Maybe note the place where `and_then` and `bind` differ from `map`
```suggestion
    ///     F: FnOnce (T) -> Option<U>;
    ///     //                          ------------ Note the return value here
```

File: src/primitive.rs [hunk 18]
```diff
@@ -0,0 +1,81 @@
+//! Primitives for constructing instances of [`IO`].
+
+use std::{ fmt::{ Display, Debug }, path::Path, fs };
+
+use super::{ IO, Inner::* };
+
+/// An alias for `pure(())`, but with a `'static` lifetime.
+#[allow(non_upper_case_globals)]
+pub const empty: IO<'static> = IO (Pure (()));
```
Comment (MaeIsBad) [inline src/primitive.rs:17]: Is this really needed? Shouldn't pure(()) produce a 'static already?

File: src/primitive.rs [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+
+/// Print the debug representation of the given value to stdout.
+pub fn debug <'a> (s: &'a impl Debug) -> IO<'a> {
+    IO::from_fn(move |cx| {
+        writeln!(cx.o, "{s:?}")
+    })
+}
+
+/// Read a file's contents into a string.
+pub fn read_file <'a> (p: impl AsRef<Path> + 'a) -> IO<'a, String> {
```
Comment (MaeIsBad) [inline src/primitive.rs]: Consider renaming this function to `read_to_string` to keep the name consistent with std, especially since reading binary files is a valid use case

**Review by MaeIsBad — COMMENTED** [review]
Seems good, a few small notes left

File: src/primitive.rs (hunk 19 above)
Comment (rlycx) [reply src/primitive.rs]: Good point. I named this function after [`readFile`](https://hackage.haskell.org/package/base-4.17.0.0/docs/Prelude.html#v:readFile) but Rust naming conventions >>>> Haskell naming conventions

File: src/primitive.rs (hunk 18 above)
Comment (rlycx) [reply src/primitive.rs:17]: yeah the docs are kinda bad here,  It's supposed to convey that it can only be `'static`. It's inspired by [`empty`](https://hackage.haskell.org/package/base-4.17.0.0/docs/Control-Applicative.html#v:empty) on `Control.Applicative.Alternative`.

Comment (rlycx) [reply src/primitive.rs:17]: Updated the comment. Thoughts?

File: src/combinator.rs [hunk 20]
```diff
@@ -0,0 +1,305 @@
...
+    /// use iom::{ IO, empty };
+    ///
+    /// # fn main () -> std::io::Result<()> {
+    /// let chain = empty.trace("I'm").trace("spamming");
+    /// let result = IO::capture(chain)?;
+    ///
+    /// assert!(result.is_empty());
+    /// # Ok (()) }
+    /// ```
+    pub fn trace (self, msg: impl Display + 'a) -> IO<'a, T> {
```
Comment (MaeIsBad) [inline src/combinator.rs:218]: Should trace require impl Debug instead?

File: src/combinator.rs [hunk 21]
```diff
@@ -0,0 +1,305 @@
...
+    /// // `b"meow"` is an array of chars, so it
+    /// // does not implement `Read`.
+    /// let buf = b"meow".as_ref();
+    /// let a = pure(buf).with_content(|s| {
+    ///     assert_eq!(s, "meow");
+    ///     empty
+    /// });
+    ///
+    /// IO::run(a)?;
+    /// # Ok (()) }
```
Comment (MaeIsBad) [inline src/combinator.rs:283]: This example should probably use assert_eq instead of a comment

**Review by MaeIsBad — APPROVED** [review]
Seems good, I haven't focused on the code and more so looked at the docs. 2 minor nitpicks but other than that merge it.
 LGBT to the moon

File: src/combinator.rs (hunk 20 above)
Comment (rlycx) [reply src/combinator.rs:218]: In this case I'd say no, since the second argument to `trace` is a message that is printed to stderr, not an object that is "traced"

File: src/combinator.rs (hunk 21 above)
Comment (rlycx) [reply src/combinator.rs:283]: I'm not sure what you mean, can you explain further?

Comment (MaeIsBad) [reply src/combinator.rs:283]: If you do
```suggestion
    /// assert!(matches!(Ok(()) = IO::run(a)?))
```

it's most in line with how rust docs tend to do things and allows doctests to check if IO::run really returns Ok(())

Comment (rlycx) [reply src/combinator.rs:283]: It needs to return `Ok (())` at the end because it is wrapped in `fn main () -> std::io::Result<()>` though. It's not a comment, rather the line with the `Ok (())` is hidden. Also, if it returns something other than `Ok (())`, the test fails due to the `?` operator, which *should* just work in tests.
